News-speaker reimplemented with changes according to request

tomascern opened the pull request:
Ahoj, zde posílám předělanou verzi news-speakeru podle požadavků Víti Nováka. Pokud by něco nebylo podle představ, můžu to ještě předělat ;)

filipmelik commented:
Hi and thanks for your contribution. Please, as a first thing - use CocoaPods package manager for the HTTP server library you used, so I can easily distinguish your files from the rest, now it's a nightmare to review to be honest. Then I will proceed with further review. 
Do not forget to add command for updating pods to the Makefile as a part of installation process and update README accordingly. 
Last thing I am not sure about - Do we need to include `sudo` directly in Makefile? What about getting rid of sudo command in Makefile and change README to instruct user to invoke `sudo make install` instead and explaining why we need a `sudo`? Test it out please and update if it works.

Thanks.

tomascern commented:
Unfortunately CocoaPods cannot be used, since this is command line tool. I wasn't able to use Http.swift as a framework neither, so I added source code into the project. But the Http.swift code is in Http and Socket folders, so you should be able to easily distinguish my files from the rest. Besides that, I made minor changes to the Http.swift library (added Server.isRunning method and fixed compiler warnings). 
And about sudo in make - I will try to fix that and commit in the afternoon.

tomascern commented:
Reorganized folder structure, my code is in speaker/Speaker directory now. Sudo was removed from Makefile and README was updated.

filipmelik commented on speaker/Speaker/main.swift:
```diff
@@ -0,0 +1,43 @@
+    
+    //check if server is running
+    if (!serverHandler.isRunning()) {
+        speaker.speak(message: "Server se nepodařilo spustit. Zkuste nastavit jiný port.")
```
Please apart from saying message out loud, add also the NSLog statement so the error will be printed to console. Otherwise, when the server is started but it could not bind to default port, it tries to read czech message with english voice ;]

Ideally, add NSLog everywhere where the error messages are being read out loud.

tomascern replied:
Added NSLogs, where applicable

filipmelik commented on speaker/Speaker/main.swift:
```diff
@@ -0,0 +1,43 @@
+import Foundation
+
+//load settings from Settings.plist
+let settings = Settings()
```
This approach is too rigid and not testable. 

1] please add possibility to pass settings file name as a daemon launch argument (positional argument is enough) To be able to launch daemon with various different configurations passed from command line. When the external argument is present, use the supplied settings file name. Otherwise use default settings file (it needs to be created, for example `defaultSettings.plist`). Inform user by NSLog message about what config file is being used.

2] Please factor out settings file loading logic from inside `Settings` object to separate class, that will load the settings into `Settings` object. Write test using XCTest framework that will verify, that the all fields in `Settings` are filled correctly by the contents from the settings plist file.

3] Please create class, that will verify, that the loaded `Settings` object is valid - i.e. that all mandatory fields are filled correctly and no one is missing. Create at least 2 test cases for invalid (incomplete) inputs and 1 for correct input.

4] When the settings file will be passed as a command line argument, I am not sure if there is any need to be able to save modified configuration file with `/set` command, because that will modify the input file itself which is imho not desired. I would leave possibility to be able to change the parameters at runtime, but I would left out the 'autosave' functionality.

tomascern replied:
1) First command line argument is now used as a path to the plist settings file. Otherwise uses defaultSettings.plist ( if not present or it is unable to parse it, create one). NSLog with information added. Just a note - to be able to launch daemon with command line argument, you need to modify com.heureka.speaker.agent.plist (add ProgramArguments item)

2) Created new class SettingsLoader.swift with loading logic from plist file into the Settings object. Added XCTestCase for testing this class.

3) In the test case, there are tests for reading valid plist file, invalid, non-existing.

4) Modifying plist file removed.

filipmelik commented on speaker/Speaker/ServerHandler.swift:
```diff
@@ -0,0 +1,183 @@
+    
+    //Set handler for '/say' function
+    private func setSayFunction() {
+        server.get("/say/{text}") { request in
```
I do not know about this server library you used, but maximum length of GET parameters are usually limited to about 2000 chars. Theroretically, when user would want to read some longer text, it would be better to handle this using POST HTTP request. 

Although I do not insist, please try to add also the POST method for reading text. Consider it as a 'bonus' task ;]

tomascern replied:
To be honest, I didn't test as long texts. Anyway, handling of POST request was added. You can try it by "curl -d "This is test"  http://localhost:8080/say"

filipmelik commented:
Two more things I've encountered:
1] Please add `copy files` phases for copying settings file and jingles to the product build directory to make it work when debugging in xcode.

2] At my machine, server quite consistently stops responding to any HTTP command after short period of time. It works correctly for about 2 minutes, but then it does not respond any more and I need to shutdown the server and start it again, to make i work for next ~2 minutes. Can you reproduce this behavior on your machine when running from xcode?

tomascern commented:
1) Copy files phases added to be able to run the application from XCode. 

2) Well, I didn't encountered this problem. Actually I was able to run the server for a whole day - including transitions into sleep mode and back.

filipmelik commented on speaker/Speaker/Settings.swift:
```diff
@@ -42,36 +55,56 @@ class Settings: NSObject {
     
     //Create dictionary with current settings to be written into plist
     private func createDictionary() -> Dictionary<String, AnyObject> {
-        return ["Volume": volume as NSNumber, "Voice": NSString(string: (voice?.rawValue ?? "")) , "Rate": rate as NSNumber, "RSS": NSString(string: (rss?.absoluteString ?? "")), "Port": rate as NSNumber, "Jingles": useJingles as NSNumber]
```
please wrap each dictionary element to the new line.

tomascern replied:
Done

filipmelik commented on speaker/Speaker/Settings.swift:
```diff
@@ -28,12 +29,24 @@ class Settings: NSObject {
         
-        let path = getPlistPath()
+        var path = ""
+        if (plistPath != nil) {
```
Although this will work and will not crash, checking for nil value like this in this context is considered unsafe. You should use optional binding here to avoid force unwrapping with `!`.

tomascern replied:
Done
